Pytest benchmark

AGaliciaMartinez opened the pull request:
# Benchamarks.

This PR contains the code that runs benchmarks on several functions. The functions that are benchmarked at this moment are:
- matmul (x20)
- add (x100)
- expm (x1)
- eigvals (x1)
The 'x20' means that the result represents 'n' operations.

The benchmarks compare the operations for different data types:
- `tensorflow` (GPU if installed, otherwise CPU)
- `numpy`
- `scipy` (some eigenvalues not implemented as was giving me some problems)
- `qt.data.Dense`
- `qt.data.CSR`

We also compare the results for when a dense or sparse input matrix is being used. Both the sparse and dense matrices are random Hermitian matrices. Sparse matrices are _tridiagonal_. The size of the employed matrix is NxN with N in {2,4,8,...,512,1024}.

# How to use it
Install the dependencies using  `pip install .[full]`
To run the benchmarks use `python benchmarks/benchmarks.py`. This will store the data and figures in the folder `.benchmarks/'.

When using `python benchmarks/benchmarks.py --help`:

usage: benchmarks.py [-h] [--save_csv SAVE_CSV] [--save_plots SAVE_PLOTS] [--plot_only]

Run and plot the benchmarks. The script also accepts the same arguments as pytest/pytest-benchmark.

optional arguments:
  -h, --help            show this help message and exit
  --save_csv SAVE_CSV   Path where the latest benchmark resulst will be stored as csv. If empty it will not store
                        results as csv. Default: .benchmarks/latest.csv
  --save_plots SAVE_PLOTS
                        Path where the plots will be saved. If empty, it will not save the plots. Default:
                        .benchmarks/figures
  --plot_only           If included, it will not run the benchmarks but just plot the latest results from .benchmaks/
                        folder.

# Examples (use `-k"expression"` to filter benchmarks")
-`python benchmarks/benchmarks.py -k"matmul-"`: Runs only the benchmarks for `matmul`.
-`python benchmarks/benchmarks.py -k"eigvals- and -dense-"`: Runs only the benchmarks for `eigvals` with dense random matrix.
-`python benchmarks/benchmarks.py -k"[512- and expm-"`: Runs only the benchmarks for `expm` for a matrix of size 512x512. (The size can only be 2,4,8...,512,1024)
-`python benchmarks/benchmarks.py -k"[2- and (tensorflow or numpy or qt.data.Dense)"`: Runs the benchmarks for every operation with matrices of size 2x2 represented with either tensorflow, numpy or qt.data.Dense.

# Example output:
Hardware: 
- CPU: intel i7-6700
- GPU: gtx-970

![add_dense](https://user-images.githubusercontent.com/57414022/122950537-d18cc080-d37c-11eb-99ab-5a27765432cf.png)
![add_sparse](https://user-images.githubusercontent.com/57414022/122950574-d81b3800-d37c-11eb-8a1d-ddf1a5adb7aa.png)
![expm_dense](https://user-images.githubusercontent.com/57414022/122950591-dd788280-d37c-11eb-8d15-e792f3603536.png)
![matmul_dense](https://user-images.githubusercontent.com/57414022/122950613-e23d3680-d37c-11eb-91d4-77074ebcd505.png)
![matmul_sparse](https://user-images.githubusercontent.com/57414022/122950620-e36e6380-d37c-11eb-8580-11b4e8e9a7f8.png)

AGaliciaMartinez committed:
Added benchmarks for matmul and add funtions.

AGaliciaMartinez committed:
Added local files to gitignore.

.asv is created by asv
.benchmarks is created by pytest-bechmarks
.ipynb_checkpoints is created by jupyter notebooks

tags* is employed to ignore the files created by ctags

AGaliciaMartinez committed:
Ignore UserWarnings.

QuTiP raises a warning if matplotlib not installed.

AGaliciaMartinez committed:
Sepparated numpy and tensorflow from qutip benchmarks.

In this way the code is more clear althoug there is some repetitions
that could be avoided.

AGaliciaMartinez committed:
Ignored imp deprecation warning.

TensorFlow raises a deprecation warning of the module imp in favor of
importlib.

AGaliciaMartinez committed:
First working version of ploting utility.

AGaliciaMartinez committed:
Moved the workflow to functions.

AGaliciaMartinez committed:
Merged tensorflows test with numpy.

AGaliciaMartinez committed:
Included tridiagonal matrix for sparse benchmarks.

The random matrix is not unitary anymore but rathe a random complex
matrix.

The sparse matrix is not density 0.1 but a tridiagonal matrix instead.

The matrices now have a seed for reprodubility.

AGaliciaMartinez committed:
Added scipy sparse and cleaned the benchmarks.

AGaliciaMartinez committed:
Disable benchmarks by default.

AGaliciaMartinez committed:
Added example test for linear algebra and excluded benchmarks from tests.

Also: Userwarning now is not being ignored.

AGaliciaMartinez committed:
A few code improvements.

AGaliciaMartinez committed:
Updated pyproject.toml

AGaliciaMartinez committed:
plot_benchmarks now runs benchmarks and acepts arguments.

The accepted arguments are those for pytest at this moment.

AGaliciaMartinez committed:
Added optional arguments to plot_benchamrks.py

AGaliciaMartinez committed:
Changed name of script.

AGaliciaMartinez committed:
Added benchmark dependencies.

AGaliciaMartinez committed:
Minor updates.

AGaliciaMartinez committed:
Renamed plot_benchmarks to benchmarks

AGaliciaMartinez commented:
I was thinking to include something like `qutip_tensorflow.benchmarks()` to execute and plot the benchmarks. This, although not necessary, would facilitate running the benchmarks in an arbitrary machine or in colab. However, since running the benchmarks requires the files in `tests`, we would require to include them in the packaging. Is this correct? Would there be a problem to include these files?

AGaliciaMartinez committed:
Removed all the filters for warnings.

AGaliciaMartinez committed:
Added expm and eigenvalues.

AGaliciaMartinez committed:
Merge branch 'main' into pytest_benchmark

quantshah commented on benchmarks/benchmarks.py:
```diff
@@ -0,0 +1,107 @@
+            d[key+'_'+k] = v
+    return data
+
+def benchmark_to_datafrane(filepath):
```
`dataframe`

AGaliciaMartinez replied:
Thank you for spotting that!

AGaliciaMartinez committed:
Correct selling mistake.

AGaliciaMartinez committed:
Added the rest of operations.

AGaliciaMartinez committed:
Removed pytest mark

AGaliciaMartinez committed:
Removed test.

AGaliciaMartinez committed:
Now mrandom matrices are hermitian.

AGaliciaMartinez committed:
Added docstrings.

AGaliciaMartinez committed:
Added docstrings.

AGaliciaMartinez committed:
Fixed bug for when B was removed as input parameter

AGaliciaMartinez committed:
Added a few comments.

AGaliciaMartinez committed:
PEP 8

AGaliciaMartinez committed:
Silenced deprecation warning.

AGaliciaMartinez commented:
The results for `expm` are a little bit surprising. Tensorflow beats the rest but only for a certain range of sizes for the matrices. For larger matrices, `numpy` seems to win. I think this has to do with the underlying function going from using a single core (below 64x64-size matrices) to using all the cores (4 cores 8 threads with intel i7-6700). It may also be something related to the cache size?

AGaliciaMartinez commented:
At this moment the test fail due to `tensorflow` raising a `DeprecationWarning` when imported due to the `imp` module being imported instead of `importlib`. Trying to silence the warning during the tests does not work due to the `-Werror ` flag for pytest. Im still thinking what to do with this.

jakelishman commented:
Cool, I will look through it properly later, but thanks for this!  I'm going to use it to test some improvements to `CSR` in general and `add_dense`.

I'm very surprised that `tensorflow` can beat `numpy` at addition.  I can only think that this isn't an apples-to-apples comparison; I think `tensorflow` must either be compiled with more vectorised CPU instructions in it, it's doing additional threaded parallelisation compared to `numpy` (can be controlled with environment variables), or it's accessed a GPU somehow?  `add` is a really simple operation at the end of the day, with easy-to-predict cache properties, so the compiled form in `numpy` should really be optimised by a compiler into the best possible instruction set for hardware.

When you import tensorflow, you can catch the warning, just like we do in QuTiP (for `imp` as well).  We have the same reason for catching it; a downstream package we have no control over imports a deprecated module.

AGaliciaMartinez commented:
The examples above use a GPU (gtx-970) for tensorflow. Given that addition is such a simple operation I was expecting tensorflow to beat numpy at leats on that one when using the gpu.

hodgestar commented on .gitignore:
```diff
@@ -22,3 +22,7 @@ develop-eggs
 .installed.cfg
 lib
 lib64
+.asv
```
What generates the `.asv` files?

AGaliciaMartinez replied:
Should not be there as we are not using asv.

hodgestar commented on .gitignore:
```diff
@@ -22,3 +22,7 @@ develop-eggs
+.asv
+.benchmarks
+.ipynb_checkpoints
+tags*
```
What generates the `tags*`?

AGaliciaMartinez replied:
`tags` is employed by [ctags](https://github.com/universal-ctags/ctags). I decided to include it here as `.vscode/` and  `/DS_Store` where also in this file although it is something probably employed only by me. Should I remove it?

hodgestar replied:
If we're not using ctags along with our code anywhere, I would leave it out.

jakelishman replied:
`tags*` is a pattern which seems like it could easily have conflicts with stuff which could be useful.  For example, it would match `tags.py` too.  At the absolute _most_ it should be `tags`/`TAGS` for ctags, but a better solution would probably be to configure your `ctags` to use a better filename, or add the file to your personal ignore files.  You can configure your local git with the variable `core.excludesFile` and point it to somewhere that has you specific files in it, or you can do it on a repository-by-repository basis by adding lines to the file `.git/info/exlude`.

Generally, the only things which should be in a repository's `.gitignore` should be files to be ignored that are actually generated as part of the build process.  However, it can often be nice to add hidden files that are really common to many editors, and have sensible, nonconflicting names, because it just eases the burden for contributors and maintainers.

hodgestar commented on benchmarks/benchmarks.py:
```diff
@@ -0,0 +1,124 @@
+
+
+def main(args=[]):
+    parser = argparse.ArgumentParser()
```
Could we perhaps add `description="...."` description of what the command does so that `--help` prints a one sentence of the script's function in addition to the parameters?

AGaliciaMartinez replied:
Indeed, I already added one but forgot to push it. Should be included now. Thank you for spotting this!

hodgestar commented on benchmarks/benchmarks.py:
```diff
@@ -0,0 +1,124 @@
+        data = pd.DataFrame(data)
+
+        # Set operation properly (for example: matmul instead of:
+        # UNSERIALIZABLE[<function Qobj.__mammal__ at 0x...)
```
```suggestion
        # UNSERIALIZABLE[<function Qobj.__matmal__ at 0x...)
```
Great typo! :D

hodgestar reviewed (approved):
I've approved, because I don't think any of my comments need to hold things up, but it would be cool if they were addressed at some point (e.g. in a future PR or small extra commits in this one -- whatever is easier).

AGaliciaMartinez committed:
Fixed typo.

AGaliciaMartinez committed:
Removed .asv from gitignore.

AGaliciaMartinez commented on benchmarks/benchmarks.py:
```diff
@@ -0,0 +1,124 @@
+def benchmark_to_dataframe(filepath):
+    """Loads a JSON file where the benchmark is stored and returns a dataframe
+    with the benchmar information."""
+    path = pathlib.Path(__file__).absolute().parents[1].joinpath(filepath)
```
I does not seem straightforward to change where the results are stored. `pytest-benchmark` accepts `--benchmark-store` but the input should be an `URI` and I do  not how to specify home directory from there. By default, it stores in './benchmarks' from wherever it is called. Also, due to how pytest test discovery works, it seems you need to execute pytest from the root the repository anyway(I'm not completely sure about this one)? 

I think for the  moment is better to assume that the user uses the script from the root of its repository.

AGaliciaMartinez committed:
Changed benchmarks to be consistent with pytest-benchmark.

It now looks and stores by default in './.benchmarks/' as
pytest-benchmark does.

AGaliciaMartinez committed:
Changed eigvals for eigvalsh

AGaliciaMartinez committed:
Removed tags from gitignore.

AGaliciaMartinez committed:
Split benchmarks into binary or unary.

AGaliciaMartinez committed:
Changed benchmark file structure.

Benchmarks are now classified as unary or binary, corresponding to the
type of operation being benchmarked. These are contained in
`benchmark_unary` and `benchmark_binary`. Operations to be benchmarked
are automatically discovered.

AGaliciaMartinez committed:
A few improvements in readability.

AGaliciaMartinez committed:
Fixed group name.

AGaliciaMartinez committed:
Fixed bug with dtype not being saved correctly.
